website/docs/proof-rules/ost.md

```diff
@@ -5,4 +5,174 @@ description: Proof rule for lower bounds on least fixpoints.
+            Body
+            tick 1
+        }
+    }
+    ```
+
+    </p>
+    </details>
+3. `I` harmonizes with `f`, i.e. they are equal on states that do not fulfill the loop guard `G`,
+    <details>
+    <summary>HeyVL Encoding</summary>
+    <p>
+
+    ```heyvl
+    proc I_harmonizes_with_post(vars: ...) -> ()
+        pre ?(!G(vars))
+        post ?(I(vars) == f(vars))
+    {}
+    ```
+
+    </p>
+    </details>
+4. the expected value of `I` after one loop iteration is finite,[^inv-iter-finite]
+    <details>
+    <summary>HeyVL Encoding</summary>
+    <p>
+
+    ```heyvl
+    @wp
+    coproc I_iter_finite(init_vars: ...) -> (vars: ...)
+        pre 0
+        post I(vars)
+    {
+        vars = init_vars // set current state to input values
+        validate // maps finite expectation to 0, ∞ to ∞
+        if G {
+            Body
+        }
+    }
+    ```
+
+    </p>
+    </details>
+5. `I` is _conditionally difference bounded_ by `c`: the absolute change of `I` by one loop iteration is bounded by `c`.
+    <details>
+    <summary>HeyVL Encoding</summary>
+    <p>
+
+    ```heyvl
+    @wp
+    coproc I_conditionally_difference_bounded_by_c(init_vars: ...) -> (vars: ...)
+        pre c
+        post ite(I(init_vars) <= I(vars), I(vars) - I(init_vars), I(init_vars) - I(vars)) // |I(vars) - I(init_vars)|
+    {
+        vars = init_vars // set current state to input values
+        Body
+    }
+    ```
+
+    </p>
+    </details>
+
+Then `I <= wp[while G { Body }](f)`.
+
+## Usage
+
+By applying the `@ost` annotation to a loop, Caesar will check the above requirements for a given invariant, PAST invariant, constant and post-expectation.
+Below is the encoding of the 'geometric loop' [Example 39 from the paper](https://dl.acm.org/doi/pdf/10.1145/3371105#page=18).
+
+```heyvl
+proc ost_geo_loop_example(init_a: Bool, init_b: UInt, init_k: UInt) -> (a: Bool, b: UInt, k: UInt)
+    pre init_b + [init_a]
+    post b
+{
+    a = init_a
+    b = init_b
+    k = init_k
+
+    @ost(
+        /* invariant */ b + [a],
+        /* past_invariant */ 2 * [a],
+        /* cdb */ 1,
+        /* post */ b
+    )
+    while a {
+        var prob_choice: Bool = flip(1 / 2)
+        if prob_choice {
+            a = false
+        } else {
+            b = b + 1
+        }
+        k = k + 1
+    }
+}
+```
+
+### Inputs
+
+You can see all four parameters are passed to the `@ost` annotation in sequence:
+
+- `invariant`: The quantitative subinvariant which should lower bound the `wp`-semantics. Must harmonize with `post`.
+- `past_invariant`: The quantitative invariant used to proof PAST.
+- `cdb`: A constant of type `UReal` specifying the conditional difference bound for the `invariant`.
+- `post`: The post-expectation after the loop.
+
+:::warning
+
+All parameters must be finite, i.e. cannot evaluate to $\infty$.
+
+:::
+
+[^1]: The built-in encoding ensures uniform integrability of the subinvariant through the conditions of the paper's Theorem 37 (b).
```

Comment:
Suggestion: "We present conditions from the paper's Theorem 37 (b) to ensure uniform integrability. One could also look into the other, alternative conditions (a) and (c) presented in the paper."